generate_figures/utilities.py

```diff
@@ -0,0 +1,55 @@
+import pandas as pd
+import wandb
+
+def plot_params(figsize, fontsize, ticksize):
+    params = {
+        "text.usetex": True,
+        "font.family": "serif",
+        "figure.figsize": figsize,
+        "figure.dpi": 80,
+        "figure.edgecolor": "k",
+        "font.size": fontsize,
+        "axes.labelsize": fontsize,
+        "axes.titlesize": fontsize,
+        "xtick.labelsize": ticksize,
+        "ytick.labelsize": ticksize,
+    }
+    return(params)
+
+
+
+def get_results_by_tag(project, tag, test_results = False):
+    api = wandb.Api()
+    count=0
+    runs = api.runs(project, filters={"tags": tag})
+    df_test_results = pd.DataFrame()
+    for i in range(len(runs)):
+        count+=1
+        df = pd.DataFrame()
+        cols = dict(runs[i].summary)
+        cols.update(dict(runs[i].config))
+        cols['id'] = runs[i].id
+        df = pd.DataFrame.from_dict(cols, orient='index').T
+ 
+        if test_results:
+            reference_path = f"{project}/run-{cols['id']}-test_results:v0"
+            test_results = api.artifact(str(reference_path)).get("test_results")
+            data = pd.DataFrame(data = test_results.data, columns = test_results.columns)
+            df = pd.concat([df, data], axis = 1)
+            df.fillna(method='ffill', inplace=True)
+        df_test_results = pd.concat([df, df_test_results])
+        
+        if count == 5:
+            return(df_test_results)  
+
+    return(df_test_results) 
+
+
+
+    # for all runs
+    # always add in the run name and id, etc.  as coulmns
+    # Then if summary = True, merge in the whole summary info
+    # if config = True, merge in the whole config 
+    # if test_resulst = False, just add that whole set of coluimns as a row
+           # else add get the test_results, then add in those columns to all of the test results then concatentate the whole dataframe.
+#Lets put this in a few clean branch off of main, add in a test_get_results.py function to the plotting which lets us play with it.
```

Comment:
There arent any type issues....I'm just gonna tell myself they changed things because they realized having them as strings is dumb because I swear I got so many type errors at the beginning.

`{'train_loss': dtype('float64'),
 'hpo_objective': dtype('O'),
 'val_u_abs_error': dtype('float64'),
 'early_stopping_monitor': dtype('O'),
 'test_loss_check_failed': dtype('bool'),
 'convergence_description': dtype('O'),
 '_step': dtype('int64'),
 'lr-Adam': dtype('float64'),
 '_runtime': dtype('float64'),
 'test_u_rel_error': dtype('O'),
 'early_stopping_check_failed': dtype('bool'),
 'epoch': dtype('int64'),
 'test_loss': dtype('O'),
 'train_time': dtype('float64'),
 'test_results': dtype('O'),
 'trainable_parameters': dtype('int64'),
 'early_stopping_threshold': dtype('float64'),
 'trainer/global_step': dtype('int64'),
 '_wandb': dtype('O'),
 'retcode': dtype('int64'),
 'X_T_mean': dtype('O'),
 'val_loss': dtype('float64'),
 '_timestamp': dtype('float64'),
 'val_u_rel_error': dtype('float64'),
 'test_u_abs_error': dtype('O'),
 'transversality_check_failed': dtype('bool'),
 'N': dtype('int64'),
 'T': dtype('int64'),
 'nu': dtype('int64'),
 'eta': dtype('float64'),
 'beta': dtype('float64'),
 'seed': dtype('int64'),
 'delta': dtype('float64'),
 'gamma': dtype('int64'),
 'sigma': dtype('float64'),
 'X_0_loc': dtype('float64'),
 'alpha_0': dtype('int64'),
 'alpha_1': dtype('int64'),
 'model.N': dtype('int64'),
 'verbose': dtype('bool'),
 'X_0_scale': dtype('float64'),
 'test_seed': dtype('int64'),
 'batch_size': dtype('int64'),
 'optimizer.lr': dtype('float64'),
 'save_metrics': dtype('bool'),
 'print_metrics': dtype('bool'),
 'shuffle_training': dtype('bool'),
 'trainer.max_time': dtype('O'),
 'val_trajectories': dtype('int64'),
 'save_test_results': dtype('bool'),
 'test_trajectories': dtype('int64'),
 'hpo_objective_name': dtype('O'),
 'train_trajectories': dtype('int64'),
 'trainer.logger.tags': dtype('O'),
 'check_transversality': dtype('bool'),
 'normalize_shock_vector': dtype('bool'),
 'omega_quadrature_nodes': dtype('int64'),
 'trainer.logger.offline': dtype('bool'),
 'always_log_hpo_objective': dtype('bool'),
 'model.ml_model.class_path': dtype('O'),
 'trainer.callbacks.monitor': dtype('O'),
 'transversality_X_mean_max': dtype('int64'),
 'transversality_X_mean_min': dtype('float64'),
 'model.ml_model.init_args.L': dtype('int64'),
 'transversality_u_rel_error': dtype('float64'),
 'test_loss_success_threshold': dtype('float64'),
 'reset_trajectories_frequency': dtype('int64'),
 'train_subsample_trajectories': dtype('int64'),
 'model.ml_model.init_args.n_in': dtype('int64'),
 'model.ml_model.init_args.n_out': dtype('int64'),
 'model.ml_model.init_args.rho_layers': dtype('int64'),
 'trainer.callbacks.stopping_threshold': dtype('float64'),
 'model.ml_model.init_args.rho_last_bias': dtype('bool'),
 'model.ml_model.init_args.rho_hidden_dim': dtype('int64'),
 'model.ml_model.init_args.rho_hidden_bias': dtype('bool'),
 'model.ml_model.init_args.rho_activator.class_path': dtype('O'),
 'id': dtype('O'),
 't': dtype('int64'),
 'ensemble': dtype('int64'),
 'u_hat': dtype('float64'),
 'residual': dtype('float64'),
 'u_reference': dtype('float64'),
 'u_rel_error': dtype('float64'),
 'u_abs_error': dtype('float64'),
 'X_min': dtype('float64'),
 'X_max': dtype('float64'),
 'X_mean': dtype('float64'),
 'X_std': dtype('float64')}`